feat: try fix

blacktoast opened the pull request:
<!-- Note: CI script will automatically rebase your PR so please do not rebase unless required -->

<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**Description:**
```

  x Type
    ,-[$DIR/tests/pass/es6/destructuring/destructuringWithLiteralInitializers2/1.ts:16:1]
 16 | function f03([x, y, z] = [1]) {}
 17 | f03();
    : ^^^
    `----

Error: 
  > ([x, y, z]?: ((number | any))[]) => void
```
Fixed an issue where the type was inferred to be an array so that it could be inferred as follows.
```
Error: 
  > [1]

  x Type
    ,-[$DIR/tests/pass/es6/destructuring/destructuringWithLiteralInitializers2/1.ts:13:1]
 13 | function f02([x, y] = [1]) {}
 14 | f02();
    : ^^^
    `----
Error: 
  > ([x, y]?: [number, any]) => void
```

I modified the `add_vars` function to do this.
Added the any type for unassigned values when declaring functions.

This fix makes a panic bug, but I'll fix it soon.

I'm wondering how to change the `any` type to `any?`

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

**BREAKING CHANGE:**

<!--
If this PR introduces a breaking change, it must contain a notice for it to be included in the CHANGELOG. Add description or remove entirely if not breaking.

You may need to update `MIGRATION.md` for the breaking changes.
-->

**Related issue (if exists):**
#720

blacktoast commented:
This is not complete, but can you please help check if I'm in the right direction?

kdy1 commented:
For any => any?, you can create a `Type::Optional`

kdy1 commented on crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:
```diff
@@ -160,20 +172,58 @@ impl Analyzer<'_, '_> {
+                        RPat::Array(p_left) => {
+                            println!("tuple@@@!!! {:?}", right);
+
+                            right = right.fold_with(&mut Widen { tuple_to_array: false });
```
The direction looks fine, but isn't this logic common for all cases?
I think this can be moved to up of the match expression

blacktoast replied:
In my opinion, `right = right.fold_with(&mut Widen { tuple_to_array: false });` is not common logic.

` right = right.fold_with(&mut Widen { tuple_to_array: ture });` is the existing logic, 
If I do this In situations like `function f03([x, y, z] = [1])`, the type is inferred as `number[]` instead of `[number]`.

So if left is an array, and `is_fn_param=true`, I want to do the logic `right = right.fold_with(&mut Widen { tuple_to_array: false });`,

kdy1 replied:
Ah, my bad. I missed the `tuple_to_array`

blacktoast commented:
I've fixed the issue #720

However, another issue has arisen.

I modified the `add_vars` function to change the type for each function parameter to fix the type inference

but The problem is :
```ts
declare const nx: undefined;
declare const sx: string;

function f33([x = 0, y = sx] = [1]) {
  y;
  x;
}
f33(); 
```

In the following code: f33() is correctly inferred to be`([x = 0, y = sx]?: [number, (string | undefined)?]) => void', but inside the function, y is also inferred as `((string | undefined)? | string)`.

To be honest, I don't know what to fix.
How can I fix it?

kdy1 commented:
You should modify the logic in `add_var`. I think adding some `dbg!` to `declare_var` will help

kdy1 commented on crates/stc_ts_type_checker/tests/tsc-stats.rust-debug:
```diff
@@ -1,6 +1,6 @@
 Stats {
```
CI failed
